packages/stryker/src/reporters/ci/Provider.ts

```diff
@@ -0,0 +1,33 @@
+import CircleProvider from './CircleProvider';
+import TravisProvider from './TravisProvider';
+
+import { getEnvironmentVariable } from '../../utils/objectUtils';
+
+/**
+ * Represents an object that can provide information about a CI/CD provider.
+ */
+export interface CiProvider {
+  /** Returns whether Stryker is running on a pull request. */
+  isPullRequest: () => boolean;
+  /** Returns the name of the Git branch of the project on which Stryker is running. */
+  determineBranch: () => string;
+  /** Returns the name of the GitHub repository of the project on which Stryker is running. */
+  determineRepository: () => string;
+}
+
+/**
+ * Return an approriate instance of CiProvider.
+ * @returns An instance of CiProvider, or undefined if it appears Stryker is not running in a CI/CD environment.
+ */
+export function determineCiProvider()  {
+  // By far the coolest env. variable from all those listed at
+  // https://docs.travis-ci.com/user/environment-variables/#Default-Environment-Variables
+  if (getEnvironmentVariable('HAS_JOSH_K_SEAL_OF_APPROVAL')) {
+    return new TravisProvider();
+  } else if (getEnvironmentVariable('CIRCLECI')) {
```

Comment:
I've considered that one, because - indeed - it would centralise knowledge of each CI system in one place. What withheld me is that it would require us to _first_ instantiate each provide and _then_ check whether it is active. But if you feel like this drawback isn't that big, it does give a nice advantage, as you pointed out.
WDYT?

Reply:
I would personally add it to the `CIProvider` interface. 

You could opt to change the `determineCiProvider` to a member of a class. In this class you can add a constructor that excepts a list of `CIProvider`s.

Something like this:

```ts
abstract class CIProvider {

  abstract isActive(): boolean;
  abstract isPullRequest(): boolean;
  abstract determineBranch(): string;
  abstract determineRepository(): string;
 
   static active( providers = [new TravisProvider(), new CircleProvider()] ): CIProvider | null {
     
   }
}
```

Reply:
Hmm all of that seems a little over-engineered in hindsight. Let's go with your implementation. We can always change it later if we want to support custom build pipelines.